python/meterpreter/ext_server_stdapi.py

```diff
@@ -2755,6 +2762,106 @@ def stdapi_ui_get_idle_time(request, response):
+    EnumDesktopsA = ctypes.windll.user32.EnumDesktopsA
+    EnumDesktopsA.argtypes = [ctypes.c_void_p, EnumDesktopCallbackPrototype, ctypes.POINTER(EnumDesktops_Param)]
+    EnumDesktopsA.restype = ctypes.c_long
+
+    WindowStationCallbackPrototype = ctypes.WINFUNCTYPE(ctypes.c_long, ctypes.c_char_p, ctypes.c_ulong)
+    EnumWindowStationsA = ctypes.windll.user32.EnumWindowStationsA
+    EnumWindowStationsA.argtypes = [WindowStationCallbackPrototype, ctypes.c_ulong]
+    EnumWindowStationsA.restype = ctypes.c_long
+
+    OpenWindowStationA = ctypes.windll.user32.OpenWindowStationA
+    OpenWindowStationA.argtypes = [ctypes.c_char_p, ctypes.c_long, ctypes.c_ulong]
+    OpenWindowStationA.restype = ctypes.c_void_p
+
+    CloseWindowStation = ctypes.windll.user32.CloseWindowStation
+    CloseWindowStation.argtypes = [ctypes.c_void_p]
+    CloseWindowStation.restype = ctypes.c_long
+
+    GetCurrentProcessId = ctypes.windll.kernel32.GetCurrentProcessId
+    GetCurrentProcessId.restype = ctypes.c_ulong
+
+    GetProcAddress = ctypes.windll.kernel32.GetProcAddress
+    GetProcAddress.argtypes = [ctypes.c_void_p, ctypes.c_char_p]
+    GetProcAddress.restype = ctypes.c_void_p
+
+    ProcessIdToSessionIdPrototype = ctypes.WINFUNCTYPE(ctypes.c_bool, ctypes.c_ulong, ctypes.POINTER(ctypes.c_ulong))
+
+    def get_session_id(pid):
+        pProcessIdToSessionId = None
+        dwSessionId = ctypes.c_ulong(0)
+
+        hKernel = ctypes.WinDLL('kernel32.dll')
+        if hKernel:
+            pProcessIdToSessionId = GetProcAddress(hKernel._handle, ctypes.c_char_p(b"ProcessIdToSessionId"))
+
+        if not pProcessIdToSessionId:
+            ctypes.windll.kernel32.FreeLibrary(None, handle=hKernel)
+            return dwSessionId
+
+        pProcessIdToSessionId = ProcessIdToSessionIdPrototype(pProcessIdToSessionId)
+        if not pProcessIdToSessionId(ctypes.c_ulong(pid), ctypes.byref(dwSessionId)):
+            dwSessionId = ctypes.c_ulong(-1)
+
+        ctypes.windll.kernel32.FreeLibrary(None, handle=hKernel)
+        return dwSessionId
+
+    @EnumDesktopCallbackPrototype
+    def desktop_enumdesktops_callback(lpszDesktop, lParam):
+        if not lParam or not lpszDesktop:
+            return True
+
+        if not lParam.contents.sessionId or not lParam.contents.cpStationName:
+            return True
+
+        entry  = bytes()
+        entry += tlv_pack(TLV_TYPE_DESKTOP_SESSION, lParam.contents.sessionId)
+        entry += tlv_pack(TLV_TYPE_DESKTOP_STATION, lParam.contents.cpStationName.decode())
+        entry += tlv_pack(TLV_TYPE_DESKTOP_NAME, lpszDesktop.decode())
+
+        global ui_desktop_enum_RESPONSE
+        ui_desktop_enum_RESPONSE += tlv_pack(TLV_TYPE_DESKTOP, entry)
+
+        return True
+
+    @WindowStationCallbackPrototype
+    def desktop_enumstations_callback(lpszWindowStation, lParam):
+        hWindowStation = OpenWindowStationA(lpszWindowStation, False, MAXIMUM_ALLOWED)
+        if not hWindowStation:
+            return False
```

Comment:
The Windows one is always returning True. If `hWindowStation` is falsy, it just breaks and doesn't call `session_id`. Returning False here will [cause the loop to terminate early.
](https://github.com/rapid7/metasploit-payloads/blob/89b469f0c92a78004aa9af0010f0b809a695b72f/c/meterpreter/source/extensions/stdapi/server/ui/desktop.c#L67)
https://github.com/rapid7/metasploit-payloads/blob/89b469f0c92a78004aa9af0010f0b809a695b72f/c/meterpreter/source/extensions/stdapi/server/ui/desktop.c#L67

It should always return True to keep the loop going.

Reply:
Nice one, fixed <3